File: python/270_ClosestBinarySearchTreeValue.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def closestValue(self, root: TreeNode, target: float) -> int:
        self.distance = [float('inf'), 0]
        def search(node):
            if not node: 
                return self.distance[1]
            
            if abs(target-node.val) < self.distance[0]:
                self.distance[0] = abs(target-node.val)
                self.distance[1] = node.val
            if target > node.val:
                return search(node.right)
            else:
                return search(node.left)
            
        return search(root)

File: cpp/270_ClosestBinarySearchTreeValue.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int closestValue(TreeNode* root, double target) {
        int closest_node = -1;
        double distance = DBL_MAX;
        
        while (root) {
            if (abs(target-root->val) < distance) {
                closest_node = root->val;
                distance = abs(target-root->val);
            }
            root = target >= root->val ? root->right : root->left;
        }
        return closest_node;
    }
};
